### PR TITLE
cabal run +RTS warning

### PR DESCRIPTION
**Pull request by BasLaa**
As mentioned in #8641, when running `cabal run +RTS`, there should be a warning that cabal consumes the `+RTS` options. This is a first test for that, still requires implementing.

### CODE REVIEW

**Comment by BasLaa** [comment]
I added a test in `cabal-testsuite/PackageTests/NewBuild/CmdRun` but am not sure that this is the correct location for it.

I am also currently having trouble getting the `cabal-install` testsuites to run alongside the `Cabal` tests. 
Using the command 'cabal-tests --with-cabal=/path/to/cabal', which path exactly should I insert? 
Also, I could not figure out how to get just one local test to run (the new one), if someone could give an example of how to do that.

**Comment by Mikolaj** [comment]
The test location and the very test looks fine to me at a quick glance.

I assume you read https://github.com/haskell/cabal/blob/master/cabal-testsuite/README.md?

In my bash history I see

    /home/mikolaj/r/cabal/dist-newstyle/build/x86_64-linux/ghc-9.0.2/cabal-testsuite-3/build/cabal-tests/cabal-tests --with-cabal /home/mikolaj/.cabal/bin/cabal  cabal-testsuite/PackageTests/NewSdist/DeterministicTrivial/deterministic.test.hs

and in general the path to cabal binary that you give is the binary with your fixes that you just compiled. You can obtain its path with `cabal list-bin cabal` after compilation.

**Comment by BasLaa** [comment]
I read the README, just couldn't figure out the right paths, turns out something wasn't added to my path properly. I'm now able to run that test locally.

**Comment by BasLaa** [comment]
I added a warning within `CmdRun.runAction` which seems to pass the new tests. It also broke quite a lot of other tests locally though. Want to see how it does here

**Review by Mikolaj — CHANGES_REQUESTED** [review]
Great job! Could you add a tiny changelog file, as well (see https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog)?

Yes, I guess the helper fits one of the Utils files.

File: cabal-install/src/Distribution/Client/CmdRun.hs
Comment (noughtmare) [inline cabal-install/src/Distribution/Client/CmdRun.hs:152]: I'm not sure if hardcoding `-N` in the warning is a good idea.

Perhaps something like this is better:

> Your RTS options are applied to cabal, not the executable.
> Use '--' to separate cabal options from your executable options.
> For example, use 'cabal run -- +RTS -N' to pass the '-N' RTS option to your executable.

I also personally think it would be good to have newlines instead of one long line, but perhaps the warning will take too much vertical space then.

**Review by noughtmare — COMMENTED** [review]
@Mikolaj you also suggested the same warning could be added to `cabal test` and `cabal  bench`. I think it would not be very hard to abstract this warning into a function and use it in multiple places. Is that still something you would like to see?

**Comment by BasLaa** [comment]
Yeah I agree with what @noughtmare says, if it is desirable for the same warning to appear with other commands, it should be no big issue to abstract this so that it can be reused.

File: cabal-install/src/Distribution/Client/CmdRun.hs
Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/CmdRun.hs:152]: I'd think the text will be wrapped by `warn` anyway. But someone needs to double-check.

Comment (BasLaa) [reply cabal-install/src/Distribution/Client/CmdRun.hs:152]: Yes, `warn` wraps text already.

**Review by Mikolaj — COMMENTED** [review]
Thank you for adding the changelog file. So what are your plans? Generalize (slightly?) already in this PR or somehow mark this as future work (e.g., open a ticket about it)?

**Comment by BasLaa** [comment]
If there's a consensus on which commands should have this error I could generalise it here? Otherwise I can push this PR as it is and create a different issue with the suggestion that other commands might want a similar error. Maybe the second approach is clearer?

**Comment by Mikolaj** [comment]
I think doing it right here for `cabal test` and `cabal bench` would let you test your abstraction in action. I don't think we need to be exhaustive --- the warning is supposed to educate the user about `--`, not catch every and all transgression. The three commands seem the most common, so it should do the job. (BTW, I've restarted failed jobs in CI to verify it's just flaky infrastructure.)

**Comment by noughtmare** [comment]
Note that for `cabal bench` you have to use `--benchmark-options="+RTS -N"' instead of just the `-- +RTS -N` separator. I don't know what works for `cabal test`.

**Comment by Mikolaj** [comment]
Well spotted. For tests it's `--test-options`. Bah, so we can't monkey-train the user so easily. :) But at least we can train the user to look for a proper option for the command in hand.

**Comment by Mikolaj** [comment]
Sounds good.

**Comment by BasLaa** [comment]
Now the warning will be given when something like `cabal bench +RTS ...` or `cabal bench -- +RTS` is run, for both `bench` and `test`.

File: cabal-install/src/Distribution/Client/CmdBench.hs
Comment (Mikolaj) [inline cabal-install/src/Distribution/Client/CmdBench.hs:105]: I'm probably misunderstanding, but does it search for a tandalone "--" in a `cabal bench` commandline? IIRC, "--" is completely ignored by `bench` and `test`, so that wouldn't make sense, would it?

Comment (BasLaa) [reply cabal-install/src/Distribution/Client/CmdBench.hs:105]: Yeah that was why I was wondering in exactly which case we want to give this warning for `bench` and `test`. The `occursOnlyOrBefore` function worked nicely for `run` as you have either `+RTS ... --` (wrong) or `-- ... +RTS` (right). With `bench` and `test`, do we want to give a warning simply when `+RTS` occurs as a standalone argument, like `cabal test +RTS`? Then this works, but I could also just make it `elem`. 

It is more difficult to check whether something like `--test-options="+RTS"` is passed, because it is treated as one argument, so you can't match the string easily. So perhaps for `bench` and `test` the best thing is just to check if there is a standalone `+RTS`. There is the edge case that a user passes something like `cabal test --test-options="+RTS" +RTS ...` and actually wants to do this, then it would still give the warning even though the user is passing RTS options to the test-suite as well.

Comment (noughtmare) [reply cabal-install/src/Distribution/Client/CmdBench.hs:105]: I think using `elem` would be the best. That the warning then also happens when users use `cabal test --test-options="+RTS" +RTS ...` is not that bad in my opinion.

Comment (Mikolaj) [reply cabal-install/src/Distribution/Client/CmdBench.hs:105]: I think warning for a standalone RTS is fine. I'd just change the warning text for test/build not to say "RTS is standalone instead of done properly", but "RTS found is standalone, which may be intentional to affect cabal, but please note RTS inside proper options is still required for RTS, if your goal is to affect the tested binary". The user may say "doh, silly computer, that's precisely why I keep the other RTS inside proper options", but I hope that's not too confusing.

Comment (BasLaa) [reply cabal-install/src/Distribution/Client/CmdBench.hs:105]: "Some RTS options were found standalone, which affect cabal instead of the binary. Please note RTS should be inside proper options if your goal is to affect the tested binary. For example, use 'cabal test --test-options="+RTS -N" to pass the '-N' RTS option to your binary."
How about this?

Comment (Mikolaj) [reply cabal-install/src/Distribution/Client/CmdBench.hs:105]: Works for me. How about "Please note RTS inside proper options suffices..." to also unconfuse a user that thinks both are needed? I'm not sure how realistic such a confusion is, but one should never underestimate users. This has the extra advantage that the user doesn't get an impression that the RTS he's put inside proper options was somehow overlooked by cabal and perhaps is invisible to cabal. In other words, we pretend we see the RTS and just make sure the *other* RTS is not put there due to the confusion that both are needed.

**Comment by BasLaa** [comment]
I think this is ready now, only the changelog needs an update

File: cabal-install/src/Distribution/Client/Utils.hs
Comment (Mikolaj) [inline cabal-install/src/Distribution/Client/Utils.hs:508]: I'd be more specific: "+RTS inside the --test-options argument". And similarly for bench.

File: cabal-install/src/Distribution/Client/Utils.hs
Comment (Mikolaj) [inline cabal-install/src/Distribution/Client/Utils.hs:510]: For silly reasons to do with shells it's in general better to do \"cabal test --test-options=\'+RTS -N'\", because the user can then use double quotes inside the options (and the other way around it doesn't work).

**Review by Mikolaj — APPROVED** [review]
Almost there IMHO.

**Comment by BasLaa** [comment]
I'm not sure why the tests are instantly failing? It seems to be something with installation, but I hardly made changes since all tests passed.

**Comment by Mikolaj** [comment]
Yes, one test finally passes, but validation tests are borked due to infrastructure problems. It may fix itself, but if not, we'll hack around that. Apologies.

**Comment by Mikolaj** [comment]
The GHA failure is hacked around; let me rebase to fix CI.

**Comment by Mikolaj** [comment]
@mergify rebase

**Comment by BasLaa** [comment]
Something still seems to be breaking in CI after merging this with the master branch

**Comment by ulysses4ever** [comment]
@BasLaa I think you didn't get #8749. Try to merge again?

**Comment by BasLaa** [comment]
It works now! The master branch keeps updating so I am merging again, then everything should be good.

**Comment by BasLaa** [comment]
@ulysses4ever good to know!

**Review by ulysses4ever — APPROVED** [review]
Thanks a lot @BasLaa!

Next step would be to apply a merge label. Either you or Mikolaj (if you don't have the access level necessary) will do it.

**Comment by andreasabel** [comment]
Thanks for the contribution!

Some merge strategy considerations:

@ulysses4ever wrote:
> Thanks a lot @BasLaa!
> Next step would be to apply a merge label. Either you or @Mikolaj (if you don't have the access level necessary) will do it.

Looks like the wrong label ("merge" instead of "squash+merge") was chosen, and we ended up with a dozen of micro-commits on `master`.

External contributors might not know the best choice here, so encouraging them to set the merge label themselves can backfire.

A technical solution would be to limit the "merge" label to administrators and only make the "squash+merge" label available to external contributors.  I don't know if this is possible, though.

Or we change our default to: _reviewer/administrator sets the merge label after ok from contributor_.

**Comment by BasLaa** [comment]
Apologies, now I know for next time!

**Comment by Mikolaj** [comment]
No big problem.

@andreasabel: Indeed, I learned to recommend "squash+merge_me" by default, unless I have the time to look at the number and kind of commits. I'd rather keep the current simple and inclusive setup, as much as the inability to git-bisect hurts us (I wonder if git can easily be persuaded to only consider merge commits). The extra ceremony of giving label permissions lets us honour contributors IMHO and them setting the label underscores their agency, which indeed is crucial given our lack of admins and maintainers.

**Comment by ulysses4ever** [comment]
Indeed, I should have been clearer, mea culpa.

I doubt you can limit which labels a contributor can apply.

**Comment by andreasabel** [comment]
Yes, it is not a big problem.  It could be a minor nuisance when cherry-picking or backporting or bisecting, otherwise it is more an aesthetic issue.

> The extra ceremony of giving label permissions lets us honour contributors IMHO and them setting the label underscores their agency,

Ok, haven't experienced it like that.  I feel "knighted" when I get the 2 approving reviews. ;-)  The actual details of the merging ceremony don't move me much; the important part is _getting_ merged.

**Comment by Mikolaj** [comment]
Oh, interesting. I wonder how the other PR authors feel. 

It's not only about feelings, though. We really do need contributors to exercise their agency and also act very responsibly, because we have a lack of admin/maintainer workpower. And we need to keep giving authors repo permissions to set labels, etc., because there's not enough active members of cabal community with said permissions to garden the repo in the long run. We celebrate a dire need and a noble duty, not only personal excellence and community camaraderie.

**Comment by ulysses4ever** [comment]
I share and support the ceremony sentiment that Mikolaj mentioned. I understand not everyone does.

**Comment by BasLaa** [comment]
Perhaps from my perspective as a recent contributor; I could have properly read through the labels and realised I should use the squash+merge label and I think it's a mistake you make just once.

It does feel good to be able to label it myself, makes you feel like you're 'in control' of your own PR.

**Pull request merged**
